### PR TITLE
Adds credentials and custom authentication support

### PR DESCRIPTION
**Pull request by yuce**
- Enables credentials and custom authentication.
- Adds `creds_username`, `creds_password` and `token_provider` configuration.
- Adds `TokenProvider` class, which is the base class for custom authentication classes.
- Adds `BasicTokenProvider` class for token-based authentication.
- Authentication docs, sample code and tests.

### CODE REVIEW

**Comment by yuce** [comment]
@mdumandag This one should be good to review.

File: .gitignore [hunk 1]
```diff
@@ -22,6 +22,7 @@ var/
 *.egg-info/
 .installed.cfg
 *.egg
+venv/
```
Comment (mdumandag) [inline .gitignore:25]: We have just added `venv/` to .gitignore. So, we could revert this change

File: docs/securing_client_connection.rst [hunk 2]
```diff
@@ -258,3 +259,72 @@ On the client side, you have to provide ``ssl_cafile``, ``ssl_certfile``
...
+        </client-permissions>
+        <member-authentication realm="passwordRealm"/>
+        <realms>
+            <realm name="passwordRealm">
+                 <identity>
+                    <username-password username="MY-USERNAME" password="MY-PASSWORD" />
+                </identity>
+            </realm>
+        </realms>
+    </security
```
Comment (mdumandag) [inline docs/securing_client_connection.rst:283]:
```suggestion
    </security>
```

File: hazelcast/token_provider.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
+
```
Comment (mdumandag) [inline hazelcast/token_provider.py]: Would it make sense to rename this module to security, so that we can put things related to security here in the future, if we want to improve the functionality we have?

Comment (mdumandag) [inline hazelcast/token_provider.py]: Unnecessary newline

File: hazelcast/token_provider.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+from six import string_types
```
Comment (mdumandag) [inline hazelcast/token_provider.py:1]: We should use `six` from `hazelcast` package

File: hazelcast/token_provider.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    def token(self):
+        # type: (BasicTokenProvider) -> bytes
+        """Returns a token to be used for token-based authentication.
+
+        Returns:
+            bytes: token as a bytes object.
+        """
+        return self._token
+
+
```
Comment (mdumandag) [inline hazelcast/token_provider.py]: Unnecessary newline

File: tests/integration/backward_compatible/authentication_tests/authentication_test.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+from hazelcast.errors import HazelcastError
+from tests.base import HazelcastTestCase
+from tests.util import get_abs_path, set_attr
+from hazelcast.client import HazelcastClient
+from hazelcast.token_provider import BasicTokenProvider
+
+
+@set_attr(enterprise=True)
+class AuthenticationTest(HazelcastTestCase):
```
Comment (mdumandag) [inline tests/integration/backward_compatible/authentication_tests/authentication_test.py:11]: We should also add this for the backward compatibility tests
```
@unittest.skipIf(
    is_client_version_older_than("4.2.1"), "Tests the features added in 4.2.1 version of the client"
)
```

File: tests/integration/backward_compatible/authentication_tests/authentication_test.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
+import os
+
+from hazelcast.errors import HazelcastError
+from tests.base import HazelcastTestCase
+from tests.util import get_abs_path, set_attr
+from hazelcast.client import HazelcastClient
+from hazelcast.token_provider import BasicTokenProvider
```
Comment (mdumandag) [inline tests/integration/backward_compatible/authentication_tests/authentication_test.py:7]: For the same reason (backward compatibility tests), we should put this import in a try-except block

File: docs/securing_client_connection.rst [hunk 8]
```diff
@@ -258,3 +259,72 @@ On the client side, you have to provide ``ssl_cafile``, ``ssl_certfile``
...
+
+You can protect your cluster using a username and password pair.
+In order to use it, enable it in member configuration:
+
+.. code:: xml
+
+    <security enabled="true">
+        <client-permissions>
+            ...
+        </client-permissions>
```
Comment (mdumandag) [inline docs/securing_client_connection.rst:274]: We can remove these unnecessary parts

File: docs/securing_client_connection.rst [hunk 9]
```diff
@@ -258,3 +259,72 @@ On the client side, you have to provide ``ssl_cafile``, ``ssl_certfile``
...
+
+In order to use token based authentication, first define in the member configuration:
+
+.. code:: xml
+
+    <security enabled="true">
+        <client-permissions>
+            ...
+        </client-permissions>
+        <member-authentication realm="tokenRealm"/>
```
Comment (mdumandag) [inline docs/securing_client_connection.rst:312]: Same comments as above

File: docs/securing_client_connection.rst [hunk 10]
```diff
@@ -258,3 +259,72 @@ On the client side, you have to provide ``ssl_cafile``, ``ssl_certfile``
...
+        <realms>
+            <realm name="tokenRealm">
+                 <identity>
+                    <token>MY-SECRET</token>
+                </identity>
+            </realm>
+        </realms>
+    </security
```
Comment (mdumandag) [inline docs/securing_client_connection.rst:320]:
```suggestion
    </security>
```

File: hazelcast/config.py [hunk 11]
```diff
@@ -11,6 +11,10 @@
     none_type,
     try_to_get_enum_value,
 )
+try:
+    from .security import TokenProvider
```
Comment (mdumandag) [inline hazelcast/config.py:16]: Why do we have such import?

File: hazelcast/config.py [hunk 12]
```diff
@@ -1290,6 +1300,36 @@ def fail_on_indeterminate_operation_state(self, value):
...
             raise TypeError("fail_on_indeterminate_operation_state must be a boolean")
 
+    @property
+    def creds_username(self):
+        # type: (_Config) -> str
+        return self._creds_username
+
+    @creds_username.setter
+    def creds_username(self, username):
+        # type: (_Config, str) -> None
```
Comment (mdumandag) [inline hazelcast/config.py:1311]: In setters, we generally make some type checks. See the others for example. There are also some unit tests we write to verify valid/invalid inputs for each config option. We should update there too.

File: docs/securing_client_connection.rst [hunk 13]
```diff
@@ -258,3 +259,72 @@ On the client side, you have to provide ``ssl_cafile``, ``ssl_certfile``
...
+
+Using :class:`hazelcast.token_provider.BasicTokenProvider` you can pass the given token the member:
+
+.. code:: python
+    token_provider = BasicTokenProvider("MY-SECRET")
+    client = hazelcast.HazelcastClient(
+        token_provider=token_provider
+    )
+
+
```
Comment (mdumandag) [inline docs/securing_client_connection.rst:330]: unnecessary newline

File: examples/authentication/authentication_example.py [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+#     </client-permissions>
+#     <member-authentication realm="passwordRealm"/>
+#     <realms>
+#         <realm name="passwordRealm">
+#              <identity>
+#                 <username-password username="member1" password="s3crEt" />
+#             </identity>
+#         </realm>
+#     </realms>
+# </security
```
Comment (mdumandag) [inline examples/authentication/authentication_example.py:25]:
```suggestion
# </security>
```

File: examples/authentication/authentication_example.py [hunk 15]
```diff
@@ -0,0 +1,35 @@
+import hazelcast
+from hazelcast.config import SSLProtocol
```
Comment (mdumandag) [inline examples/authentication/authentication_example.py:2]: unused import

File: .gitignore (hunk 1 above)
Comment (yuce) [reply .gitignore]: OK: https://github.com/hazelcast/hazelcast-python-client/pull/446/commits/ebb63d27ca7bbabd044553f68e3606bd43cbb426

File: docs/securing_client_connection.rst (hunk 2 above)
Comment (yuce) [reply docs/securing_client_connection.rst]: OK: https://github.com/hazelcast/hazelcast-python-client/pull/446/commits/36946e76b40c9e937d31840499db6bd23f1b14c6

File: hazelcast/token_provider.py (hunk 3 above)
Comment (yuce) [reply hazelcast/token_provider.py]: OK: https://github.com/hazelcast/hazelcast-python-client/pull/446/commits/36946e76b40c9e937d31840499db6bd23f1b14c6

File: hazelcast/token_provider.py (hunk 4 above)
Comment (yuce) [reply hazelcast/token_provider.py]: OK: https://github.com/hazelcast/hazelcast-python-client/pull/446/commits/36946e76b40c9e937d31840499db6bd23f1b14c6

File: tests/integration/backward_compatible/authentication_tests/authentication_test.py (hunk 7 above)
Comment (yuce) [reply tests/integration/backward_compatible/authentication_tests/authentication_test.py]: OK: https://github.com/hazelcast/hazelcast-python-client/pull/446/commits/36946e76b40c9e937d31840499db6bd23f1b14c6

File: tests/integration/backward_compatible/authentication_tests/authentication_test.py (hunk 6 above)
Comment (yuce) [reply tests/integration/backward_compatible/authentication_tests/authentication_test.py:19]: OK: https://github.com/hazelcast/hazelcast-python-client/pull/446/commits/36946e76b40c9e937d31840499db6bd23f1b14c6 and https://github.com/hazelcast/hazelcast-python-client/pull/446/commits/16b71978621c5fca41e3ed84810d563834847c91

File: examples/authentication/authentication_example.py (hunk 15 above)
Comment (yuce) [reply examples/authentication/authentication_example.py]: OK: https://github.com/hazelcast/hazelcast-python-client/pull/446/commits/36946e76b40c9e937d31840499db6bd23f1b14c6

File: examples/authentication/authentication_example.py (hunk 14 above)
Comment (yuce) [reply examples/authentication/authentication_example.py]: OK: https://github.com/hazelcast/hazelcast-python-client/pull/446/commits/36946e76b40c9e937d31840499db6bd23f1b14c6

File: docs/securing_client_connection.rst (hunk 13 above)
Comment (yuce) [reply docs/securing_client_connection.rst]: OK: https://github.com/hazelcast/hazelcast-python-client/pull/446/commits/28c43f3a55de4be71cd83c366eb4fd8062f152e8

File: hazelcast/config.py (hunk 12 above)
Comment (yuce) [reply hazelcast/config.py:1308]: OK: https://github.com/hazelcast/hazelcast-python-client/pull/446/commits/36946e76b40c9e937d31840499db6bd23f1b14c6

File: hazelcast/config.py (hunk 11 above)
Comment (yuce) [reply hazelcast/config.py]: Importing `TokenProvider` is necessary for the type annotation. The `try` is a leftover. Fixed in: https://github.com/hazelcast/hazelcast-python-client/pull/446/commits/36946e76b40c9e937d31840499db6bd23f1b14c6

File: docs/securing_client_connection.rst (hunk 10 above)
Comment (yuce) [reply docs/securing_client_connection.rst]: OK: https://github.com/hazelcast/hazelcast-python-client/pull/446/commits/36946e76b40c9e937d31840499db6bd23f1b14c6

File: docs/securing_client_connection.rst (hunk 8 above)
Comment (yuce) [reply docs/securing_client_connection.rst]: OK: https://github.com/hazelcast/hazelcast-python-client/pull/446/commits/1b5f023e268221a1f09d54c8b47e1adbde7e57d1

File: docs/securing_client_connection.rst (hunk 9 above)
Comment (yuce) [reply docs/securing_client_connection.rst]: OK: https://github.com/hazelcast/hazelcast-python-client/pull/446/commits/1b5f023e268221a1f09d54c8b47e1adbde7e57d1

File: docs/securing_client_connection.rst [hunk 16]
```diff
@@ -258,3 +259,67 @@ On the client side, you have to provide ``ssl_cafile``, ``ssl_certfile``
...
+A token provider is a class derived from :class:`hazelcast.security.TokenProvider`.
+
+In order to use token based authentication, first define in the member configuration:
+
+.. code:: xml
+
+    <security enabled="true">
+        <client-permissions>
+            ...
+        </client-permissions>
```
Comment (mdumandag) [inline docs/securing_client_connection.rst:308]: I see that these 3 lines added back, probably mistakenly

File: examples/security/token_authentication_example.py [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+#     </client-permissions>
+#     <member-authentication realm="tokenRealm"/>
+#     <realms>
+#         <realm name="tokenRealm">
+#              <identity>
+#                 <token>s3crEt</token>
+#             </identity>
+#         </realm>
+#     </realms>
+# </security
```
Comment (mdumandag) [inline examples/security/token_authentication_example.py:25]:
```suggestion
# </security>
```

File: examples/security/username_password_authentication_example.py [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+#     </client-permissions>
+#     <member-authentication realm="passwordRealm"/>
+#     <realms>
+#         <realm name="passwordRealm">
+#              <identity>
+#                 <username-password username="member1" password="s3crEt" />
+#             </identity>
+#         </realm>
+#     </realms>
+# </security
```
Comment (mdumandag) [inline examples/security/username_password_authentication_example.py:24]:
```suggestion
# </security>
```

File: hazelcast/config.py [hunk 19]
```diff
@@ -1290,6 +1298,43 @@ def fail_on_indeterminate_operation_state(self, value):
...
+    def token_provider(self):
+        # type: (_Config) -> TokenProvider
+        return self._token_provider
+
+    @token_provider.setter
+    def token_provider(self, token_provider):
+        # type: (_Config, TokenProvider) -> None
+        token_fun = getattr(token_provider, "token", None)
+        if token_fun is None or not isinstance(token_fun, types.MethodType):
+            raise TypeError("token_provider must be an object with a token method")
```
Comment (mdumandag) [inline hazelcast/config.py:1335]: I see that we accept anything with a `token()` method but I feel like we should make use of the `TokenProvider` class.

Something like
```suggestion
        if not isinstance(token_provider, TokenProvider):
            raise TypeError("token_provider must be an instance of TokenProvider")
```

File: hazelcast/connection.py [hunk 20]
```diff
@@ -516,10 +531,7 @@ def _on_auth(self, response, connection):
             return self._handle_successful_auth(response, connection)
 
         if status == _AuthenticationStatus.CREDENTIALS_FAILED:
-            err = AuthenticationError(
-                "Authentication failed. The configured cluster name on "
-                "the client does not match the one configured in the cluster."
-            )
+            err = AuthenticationError("Authentication failed. Check cluster name and credentials.")
```
Comment (mdumandag) [inline hazelcast/connection.py:534]: Maybe make the error message better with something like this.

```suggestion
            err = AuthenticationError("Authentication failed. Either the configured cluster name on the client does not match with the one configured on the cluster or the credentials set in the client could not be authenticated.")
```
(Please make line shorter by line breaks if you are going to make this change)

File: hazelcast/security/token_provider.py [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+class BasicTokenProvider(TokenProvider):
+    """BasicTokenProvider sends the given token to the authentication endpoint."""
+
+    def __init__(self, token=""):
+        if isinstance(token, string_types):
+            self._token = token.encode("utf-8")
+        elif isinstance(token, bytes):
+            self._token = token
+        else:
+            raise ValueError("token must be either a str or bytes object")
```
Comment (mdumandag) [inline hazelcast/security/token_provider.py:26]: This is more like a `TypeError`

File: tests/integration/backward_compatible/authentication_tests/authentication_test.py [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+
+try:
+    from hazelcast.security import BasicTokenProvider
+except ImportError:
+    pass
+
+
+@set_attr(enterprise=True)
+@unittest.skipIf(
+    is_client_version_older_than("4.2.2"), "Tests the features added in 4.2.2 version of the client"
```
Comment (mdumandag) [inline tests/integration/backward_compatible/authentication_tests/authentication_test.py:17]: Why `4.2.2`? We are about to release `4.2.1`

File: hazelcast/config.py (hunk 19 above)
Comment (yuce) [reply hazelcast/config.py:1335]: Since Kerberos token provider is in its own package, that creates a chicken-egg problem ATM. We can do that after releasing 4.2.2.

File: hazelcast/connection.py (hunk 20 above)
Comment (yuce) [reply hazelcast/connection.py:534]: Shorter one conveys the same information unless I am missing anything.

File: hazelcast/config.py (hunk 19 above)
Comment (yuce) [reply hazelcast/config.py:1335]: ^ I mean 4.2.1

File: tests/integration/backward_compatible/authentication_tests/authentication_test.py (hunk 22 above)
Comment (yuce) [reply tests/integration/backward_compatible/authentication_tests/authentication_test.py]: Don't know why I was thinking we are about to release 4.2.2. Fixed: https://github.com/hazelcast/hazelcast-python-client/pull/446/commits/1e68b2525e4de8cae7891a7fe90ac287d16d0173

File: hazelcast/security/token_provider.py (hunk 21 above)
Comment (yuce) [reply hazelcast/security/token_provider.py]: OK: https://github.com/hazelcast/hazelcast-python-client/pull/446/commits/1e68b2525e4de8cae7891a7fe90ac287d16d0173

File: examples/security/username_password_authentication_example.py (hunk 18 above)
Comment (yuce) [reply examples/security/username_password_authentication_example.py]: OK: https://github.com/hazelcast/hazelcast-python-client/pull/446/commits/1e68b2525e4de8cae7891a7fe90ac287d16d0173

File: examples/security/token_authentication_example.py (hunk 17 above)
Comment (yuce) [reply examples/security/token_authentication_example.py]: OK: https://github.com/hazelcast/hazelcast-python-client/pull/446/commits/1e68b2525e4de8cae7891a7fe90ac287d16d0173

File: docs/securing_client_connection.rst (hunk 16 above)
Comment (yuce) [reply docs/securing_client_connection.rst]: OK: https://github.com/hazelcast/hazelcast-python-client/pull/446/commits/1e68b2525e4de8cae7891a7fe90ac287d16d0173

File: hazelcast/config.py (hunk 19 above)
Comment (mdumandag) [reply hazelcast/config.py:1335]: Won't the Kerberos package depend on the Hazelcast client 4.2.1+?

Comment (yuce) [reply hazelcast/config.py:1335]: It doesn't ATM. IMO there's not much need to add a dependency but we may discuss that after we release 4.2.1.

File: hazelcast/connection.py (hunk 20 above)
Comment (mdumandag) [reply hazelcast/connection.py:534]: Yes, they are more or less the same but I feel like the longer one expresses the situation better.
The authentication failed -> Here are the causes
Of course, you can say that the shorter version might say the same thing.
I am resolving this, feel free to ignore this

File: hazelcast/config.py (hunk 19 above)
Comment (mdumandag) [reply hazelcast/config.py:1335]: OK. BTW, where is the code for it? I haven't seen it

**Comment by mdumandag** [comment]
I am OK with the code now. I will approve once the tests on Jenkins pass

**Review by mdumandag — APPROVED** [review]
